File: haskell/ch1/Ex128.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...
module Ex128 where

import System.Random
import Control.Monad
import Data.Char


fast_times = 32

sq n = n * n

divides a b = (mod b a) == 0

get_rng seed = mkStdGen seed

rand_int :: Int -> StdGen -> (Int, StdGen)
rand_int limit gen =
  let (rawTargetNum, nextGen) = next gen
  in (rem rawTargetNum limit, nextGen)
 
find_divisor n test | sq test > n = n
find_divisor n test | divides test n = test
find_divisor n test = find_divisor n (test + 1)

expmod :: Int -> Int -> Int -> Int
expmod _ 0 _ = 1
expmod base exp m | even exp = rem (sq (expmod base (div exp 2) m)) m
expmod base exp m = rem (base * (expmod base (exp - 1) m)) m

non_trivial_square_root :: Int -> Int -> Int -> Int
non_trivial_square_root x square modulus =
  if and [square == 1,
          not (x == 1),
          not (x == (modulus-1))]
  then 0
  else square
                                     
miller_rabin_expmod :: Int -> Int -> Int -> Int
miller_rabin_expmod _ 0 _ = 1
miller_rabin_expmod base exp m | even exp =
                                 let squaremod_with_check x =
                                       non_trivial_square_root x (rem (sq x) m) m
                                 in
                                   squaremod_with_check (miller_rabin_expmod base (div exp 2) m)
miller_rabin_expmod base exp m = rem (base * (miller_rabin_expmod base (exp - 1) m)) m

fermat_test :: Int -> Int -> Bool
fermat_test n rnum = 
  let try_it a = (expmod a n n) == a
  in try_it (1 + rnum)

miller_rabin_test :: Int -> Int -> Bool
miller_rabin_test n rnum = 
  let try_it a = 1 == miller_rabin_expmod a (n-1) n
  in try_it (1 + rnum)

fast_prime n 0 _ = True
fast_prime n times gen =
  let (rnum, newGen) = rand_int (n-1) gen
  in
  if fermat_test n rnum 
  then fast_prime n (times - 1) newGen
  else False

fast_miller_rabin n 0 _ = True
fast_miller_rabin n times gen =
  let (rnum, newGen) = rand_int (n-1) gen
  in
  if miller_rabin_test n rnum 
  then fast_miller_rabin n (times - 1) newGen
  else False

--is_prime n = fast_prime n fast_times (get_rng n)
is_prime n = (find_divisor n 2) == n

test_num n =
  (n,
   is_prime n,
   fast_prime n 128 (get_rng n),
   fast_miller_rabin n 128 (get_rng n)
  )

test_nums vals = report $ map show (map test_num vals)

report vals = mapM_ putStrLn vals


main = do
  putStrLn "(n,Prime,Fast Prime,Miller Rabin)"
  test_nums [11,13,127,561,1105,1729,2465,2821,6601,6603]

File: python/ch1/ex128.py
import unittest
import random

FAST_TIMES = 4


def divides(a, b):
    return b % a == 0


def square(n):
    return n * n


def smallest_divisor(n):
    return find_divisor(n, 2)


def find_divisor(n, test):
    if square(test) > n:
        return n
    elif divides(test, n):
        return test
    else:
        return find_divisor(n, test+1)


def prime(n):
    return n == smallest_divisor(n)


def fast_prime(n, times):
    if times == 0:
        return True
    if fermat_test(n):
        return fast_prime(n, times - 1)
    return False


def fermat_test(n):
    def try_it(a):
        return a == expmod(a, n, n)
    return try_it(random.randint(1, n))


def expmod(base, exp, m):
    if exp == 0:
        return 1
    elif divides(2, exp):
        return square(expmod(base, exp/2, m)) % m
    else:
        return (base * expmod(base, exp-1, m)) % m


def fast_miller_rabin(n, times):
    if times == 0:
        return True
    if miller_rabin_test(n):
        return fast_miller_rabin(n, times - 1)
    return False


def miller_rabin_test(n):
    def try_it(a):
        return 1 == miller_rabin_expmod(a, n-1, n)
    return try_it(random.randint(1, n))


def miller_rabin_expmod(base, exp, m):
    def non_trivial_square_root(x, sq):
        if sq==1 and x!=1 and x!=(m-1):
            return 0
        else:
            return sq

    def squaremod_with_check(x):
        return non_trivial_square_root(x, square(x) % m)

    if exp == 0:
        return 1
    elif divides(2, exp):
        return squaremod_with_check(miller_rabin_expmod(base, exp/2, m))
    else:
        return (base * miller_rabin_expmod(base, exp-1, m)) % m


def test_number(n):
    result = (n, prime(n), fast_prime(n, FAST_TIMES), fast_miller_rabin(n, FAST_TIMES))
    print("%d \t Prime: %s \t Fast Prime: %s \t Miller-Rabin: %s" % result)


class MyTestCase(unittest.TestCase):
    def test2(self):
        test_number(10)
        test_number(81)
        test_number(6603)
        test_number(23)
        test_number(41)
        test_number(179)
        test_number(503)
        test_number(561)
        test_number(1105)
        test_number(1729)
        test_number(2465)
        test_number(2821)
        test_number(6601)


if __name__ == '__main__':
    unittest.main()
